### ISSUE TITLE
Error: Multibyte text in HTTP request

### ISSUE THREAD

**Issue by vermiculus** [tarsius/ghub#12]
Reproduce with:

```elisp
(ghub-post "/markdown" nil '((text . "‘guess’")))
```
Note the special quotes around `guess`. This particular text is copied from a help-mode buffer.
```
Debugger entered--Lisp error: (error "Multibyte text in HTTP request: POST /markdown HTTP/1.1
MIME-Version: 1.0
Connection: keep-alive
Extension: Security/Digest Security/SSL
Host: api.github.com
Accept-encoding: gzip
Accept: */*
User-Agent: URL/Emacs
Cookie: HttpOnly=nil; logged_in=no
Content-Type: application/json
Authorization: token ac3ee799f0108409d026deeb1bb51892e9b4c8de
Content-length: 18

{\"text\":\"‘guess’\"}")
  signal(error ("Multibyte text in HTTP request: POST /markdown HTTP/1.1
\nMIME-Version: 1.0
\nConnection: keep-alive
\nExtension: Security/Digest Security/SSL
\nHost: api.github.com
\nAccept-encoding: gzip
\nAccept: */*
\nUser-Agent: URL/Emacs
\nCookie: HttpOnly=nil; logged_in=no
\nContent-Type: application/json
\nAuthorization: token ac3ee799f0108409d026deeb1bb51892e9b4c8de
\nContent-length: 18
\n
\n{\"text\":\"‘guess’\"}"))
  error("Multibyte text in HTTP request: %s" "POST /markdown HTTP/1.1
\nMIME-Version: 1.0
\nConnection: keep-alive
\nExtension: Security/Digest Security/SSL
\nHost: api.github.com
\nAccept-encoding: gzip
\nAccept: */*
\nUser-Agent: URL/Emacs
\nCookie: HttpOnly=nil; logged_in=no
\nContent-Type: application/json
\nAuthorization: token ac3ee799f0108409d026deeb1bb51892e9b4c8de
\nContent-length: 18
\n
\n{\"text\":\"‘guess’\"}")
  url-http-create-request()
  url-http([cl-struct-url "https" nil nil "api.github.com" nil "/markdown" nil nil t nil t] #[128 "\302\303\304p#\210\300\305\240\210\301p\240\207" [(nil) (nil) url-debug retrieval "Synchronous fetching done (%S)" t] 5 "\n\n(fn &rest IGNORED)"] (nil) nil tls)
  url-https([cl-struct-url "https" nil nil "api.github.com" nil "/markdown" nil nil t nil t] #[128 "\302\303\304p#\210\300\305\240\210\301p\240\207" [(nil) (nil) url-debug retrieval "Synchronous fetching done (%S)" t] 5 "\n\n(fn &rest IGNORED)"] (nil))
  url-retrieve-internal("https://api.github.com/markdown" #[128 "\302\303\304p#\210\300\305\240\210\301p\240\207" [(nil) (nil) url-debug retrieval "Synchronous fetching done (%S)" t] 5 "\n\n(fn &rest IGNORED)"] (nil) nil nil)
  url-retrieve("https://api.github.com/markdown" #[128 "\302\303\304p#\210\300\305\240\210\301p\240\207" [(nil) (nil) url-debug retrieval "Synchronous fetching done (%S)" t] 5 "\n\n(fn &rest IGNORED)"] nil nil nil)
  url-retrieve-synchronously("https://api.github.com/markdown")
  (set-buffer (url-retrieve-synchronously (concat ghub-base-url resource p)))
  (save-current-buffer (set-buffer (url-retrieve-synchronously (concat ghub-base-url resource p))) (set-buffer-multibyte t) (let (link body) (goto-char (point-min)) (save-restriction (narrow-to-region (point) url-http-end-of-headers) (and (setq link (mail-fetch-field "Link")) (setq link (car (rassoc (list "rel=\"next\"") (mapcar ... ...)))) (string-match "[?&]page=\\([^&>]+\\)" link) (setq link (match-string 1 link)))) (goto-char (1+ url-http-end-of-headers)) (setq body (ghub--read-response)) (if (or noerror (= (/ url-http-response-status 100) 2)) nil (cond ((eq url-http-response-status 301) (let nil (signal (quote ghub-301) (list method resource p d body)))) ((eq url-http-response-status 400) (let nil (signal (quote ghub-400) (list method resource p d body)))) ((eq url-http-response-status 404) (let nil (signal (quote ghub-404) (list method resource p d body)))) ((eq url-http-response-status 422) (let nil (signal (quote ghub-422) (list method resource p d body)))) (t (let nil (signal (quote ghub-http-error) (list url-http-response-status method resource p d body)))))) (if (and link ghub-unpaginate) (nconc body (ghub--request method resource (cons (cons (quote page) link) (cl-delete (quote page) params :key (function car))) data noerror)) body)))
  (let* ((p (and params (concat "?" (ghub--url-encode-params params)))) (d (and data (json-encode-list data))) (url-request-extra-headers (cons (quote ("Content-Type" . "application/json")) (let* ((token (and t ...))) (if token (list (cons "Authorization" ...)))))) (url-request-method method) (url-request-data d)) (save-current-buffer (set-buffer (url-retrieve-synchronously (concat ghub-base-url resource p))) (set-buffer-multibyte t) (let (link body) (goto-char (point-min)) (save-restriction (narrow-to-region (point) url-http-end-of-headers) (and (setq link (mail-fetch-field "Link")) (setq link (car (rassoc ... ...))) (string-match "[?&]page=\\([^&>]+\\)" link) (setq link (match-string 1 link)))) (goto-char (1+ url-http-end-of-headers)) (setq body (ghub--read-response)) (if (or noerror (= (/ url-http-response-status 100) 2)) nil (cond ((eq url-http-response-status 301) (let nil (signal ... ...))) ((eq url-http-response-status 400) (let nil (signal ... ...))) ((eq url-http-response-status 404) (let nil (signal ... ...))) ((eq url-http-response-status 422) (let nil (signal ... ...))) (t (let nil (signal ... ...))))) (if (and link ghub-unpaginate) (nconc body (ghub--request method resource (cons (cons ... link) (cl-delete ... params :key ...)) data noerror)) body))))
  (closure (url-http-end-of-headers url-http-response-status t) (method resource &optional params data noerror) (let* ((p (and params (concat "?" (ghub--url-encode-params params)))) (d (and data (json-encode-list data))) (url-request-extra-headers (cons (quote ("Content-Type" . "application/json")) (let* ((token ...)) (if token (list ...))))) (url-request-method method) (url-request-data d)) (save-current-buffer (set-buffer (url-retrieve-synchronously (concat ghub-base-url resource p))) (set-buffer-multibyte t) (let (link body) (goto-char (point-min)) (save-restriction (narrow-to-region (point) url-http-end-of-headers) (and (setq link (mail-fetch-field "Link")) (setq link (car ...)) (string-match "[?&]page=\\([^&>]+\\)" link) (setq link (match-string 1 link)))) (goto-char (1+ url-http-end-of-headers)) (setq body (ghub--read-response)) (if (or noerror (= (/ url-http-response-status 100) 2)) nil (cond ((eq url-http-response-status 301) (let nil ...)) ((eq url-http-response-status 400) (let nil ...)) ((eq url-http-response-status 404) (let nil ...)) ((eq url-http-response-status 422) (let nil ...)) (t (let nil ...)))) (if (and link ghub-unpaginate) (nconc body (ghub--request method resource (cons ... ...) data noerror)) body)))))("POST" "/markdown" nil ((text . "‘guess’")) nil)
  apply((closure (url-http-end-of-headers url-http-response-status t) (method resource &optional params data noerror) (let* ((p (and params (concat "?" (ghub--url-encode-params params)))) (d (and data (json-encode-list data))) (url-request-extra-headers (cons (quote ("Content-Type" . "application/json")) (let* ((token ...)) (if token (list ...))))) (url-request-method method) (url-request-data d)) (save-current-buffer (set-buffer (url-retrieve-synchronously (concat ghub-base-url resource p))) (set-buffer-multibyte t) (let (link body) (goto-char (point-min)) (save-restriction (narrow-to-region (point) url-http-end-of-headers) (and (setq link (mail-fetch-field "Link")) (setq link (car ...)) (string-match "[?&]page=\\([^&>]+\\)" link) (setq link (match-string 1 link)))) (goto-char (1+ url-http-end-of-headers)) (setq body (ghub--read-response)) (if (or noerror (= (/ url-http-response-status 100) 2)) nil (cond ((eq url-http-response-status 301) (let nil ...)) ((eq url-http-response-status 400) (let nil ...)) ((eq url-http-response-status 404) (let nil ...)) ((eq url-http-response-status 422) (let nil ...)) (t (let nil ...)))) (if (and link ghub-unpaginate) (nconc body (ghub--request method resource (cons ... ...) data noerror)) body))))) ("POST" "/markdown" nil ((text . "‘guess’")) nil))
  (if (and (listp magithub-debug-mode) (memq (quote dry-api) magithub-debug-mode)) nil (apply oldfun args))
  magithub-debug--ghub-request-wrapper((closure (url-http-end-of-headers url-http-response-status t) (method resource &optional params data noerror) (let* ((p (and params (concat "?" (ghub--url-encode-params params)))) (d (and data (json-encode-list data))) (url-request-extra-headers (cons (quote ("Content-Type" . "application/json")) (let* ((token ...)) (if token (list ...))))) (url-request-method method) (url-request-data d)) (save-current-buffer (set-buffer (url-retrieve-synchronously (concat ghub-base-url resource p))) (set-buffer-multibyte t) (let (link body) (goto-char (point-min)) (save-restriction (narrow-to-region (point) url-http-end-of-headers) (and (setq link (mail-fetch-field "Link")) (setq link (car ...)) (string-match "[?&]page=\\([^&>]+\\)" link) (setq link (match-string 1 link)))) (goto-char (1+ url-http-end-of-headers)) (setq body (ghub--read-response)) (if (or noerror (= (/ url-http-response-status 100) 2)) nil (cond ((eq url-http-response-status 301) (let nil ...)) ((eq url-http-response-status 400) (let nil ...)) ((eq url-http-response-status 404) (let nil ...)) ((eq url-http-response-status 422) (let nil ...)) (t (let nil ...)))) (if (and link ghub-unpaginate) (nconc body (ghub--request method resource (cons ... ...) data noerror)) body))))) "POST" "/markdown" nil ((text . "‘guess’")) nil)
  apply(magithub-debug--ghub-request-wrapper (closure (url-http-end-of-headers url-http-response-status t) (method resource &optional params data noerror) (let* ((p (and params (concat "?" (ghub--url-encode-params params)))) (d (and data (json-encode-list data))) (url-request-extra-headers (cons (quote ("Content-Type" . "application/json")) (let* ((token ...)) (if token (list ...))))) (url-request-method method) (url-request-data d)) (save-current-buffer (set-buffer (url-retrieve-synchronously (concat ghub-base-url resource p))) (set-buffer-multibyte t) (let (link body) (goto-char (point-min)) (save-restriction (narrow-to-region (point) url-http-end-of-headers) (and (setq link (mail-fetch-field "Link")) (setq link (car ...)) (string-match "[?&]page=\\([^&>]+\\)" link) (setq link (match-string 1 link)))) (goto-char (1+ url-http-end-of-headers)) (setq body (ghub--read-response)) (if (or noerror (= (/ url-http-response-status 100) 2)) nil (cond ((eq url-http-response-status 301) (let nil ...)) ((eq url-http-response-status 400) (let nil ...)) ((eq url-http-response-status 404) (let nil ...)) ((eq url-http-response-status 422) (let nil ...)) (t (let nil ...)))) (if (and link ghub-unpaginate) (nconc body (ghub--request method resource (cons ... ...) data noerror)) body))))) ("POST" "/markdown" nil ((text . "‘guess’")) nil))
  ghub--request("POST" "/markdown" nil ((text . "‘guess’")) nil)
  ghub-post("/markdown" nil ((text . "‘guess’")))
```

**Comment by tarsius** [comment]
Sounds like `url` has to be taught to (optionally) accept multibyte text. Maybe there already is a mechanism to allow this. Could you please check?

**Comment by tarsius** [comment]
This gives the same error:

```lisp
(ghub-post "/markdown" nil `((text . ,(string-as-unibyte "‘guess’"))))
```

 even though in this case the last line of the error/response is

```
{"text":"\342\200\230guess\342\200\231"}
```

If you want to see this addressed soon, then you will have to ask someone who understands encodings better than us. I have no idea how to make this work.

**Comment by tarsius** [comment]
Also see https://debbugs.gnu.org/cgi/bugreport.cgi?bug=23750.

**Comment by tarsius** [comment]
For testing without using `ghub`:

```lisp
(let ((url-request-extra-headers
       `(("Content-Type"  . "application/json")))
      (url-request-method "POST")
      (url-request-data
       (json-encode-list
        (list (cons 'text
                    (identity "‘guess’"))))))
  (with-current-buffer
      (url-retrieve-synchronously "https://api.github.com/markdown")
    (buffer-substring-no-properties (1+ url-http-end-of-headers) (point-max))))
```

Hopefully there is something that we can use to replace `identity` that makes this work.

**Comment by tarsius** [comment]
@dgutov Could you please help us with this? In https://debbugs.gnu.org/cgi/bugreport.cgi?bug=23750#20 you said:

> While the truncation bug itself is quite old, it's been exacerbated in 
> Emacs 25 by my own цщкл to make json.el faster: one side-effect is  that 
> it doesn't \u-quote multibyte characters anymore, or at least not all of 
> them.

Is there a way to undo that when needed, as it appears to be in our case?

**Comment by dgutov** [comment]
You're only allowed to send unibyte strings (because the lower level, strictly speaking, doesn't know which encoding to use).

So please use `encode-coding-string`, like mentioned in https://debbugs.gnu.org/cgi/bugreport.cgi?bug=23750#29.

**Comment by tarsius** [comment]
I really know very little about all this. Could please show me the correct way of doing this by modifying the above code snippet?

**Comment by dgutov** [comment]
Here you go:

```
(let ((url-request-extra-headers
       `(("Content-Type"  . "application/json")))
      (url-request-method "POST")
      (url-request-data
       (encode-coding-string
        (json-encode-list
         (list (cons 'text
                     (identity "‘guess’"))))
        'utf-8)))
  (with-current-buffer
      (url-retrieve-synchronously "https://api.github.com/markdown")
    (buffer-substring-no-properties (1+ url-http-end-of-headers) (point-max))))
```

**Comment by tarsius** [comment]
Thanks!

I also had to `(decode-coding-string ... 'utf-8)` to get back `"<p>‘guess’</p>\n"` instead of
 `"<p>\342\200\230guess\342\200\231</p>\n"`.

I would never have guessed that I am supposed to use `utf-8`. Aren't we doing all this because multibyte strings are *not* supported (`url-http-create-request: Multibyte text in HTTP request`) * and isn't `utf-8` "multibyte"? What was used before `utf-8` was specified explicilty?

(*) Or did I misunderstand this and it means "Unless specified otherwise unibyte text is assumed, but there is multibye text in http request anyway"?

**Comment by dgutov** [comment]
> I would never have guessed that I am supposed to use utf-8

Which other encoding do you want to make the HTTP request in? Use it, then. Maybe also add an appropriate HTTP header, so the server on the other side can decode it.

> Aren't we doing all this because multibyte strings are not supported * and isn't utf-8 "multibyte"?

encode-coding-string returns a unibyte string. A byte sequence, if you will.

> What was used before utf-8 was specified explicilty?

It was being done implicitly, by accident. With an accompanying bug (or at least one that we know of).

> I also had to (decode-coding-string ... 'utf-8)

Not sure about that, to be honest.

**Comment by tarsius** [comment]
(I'll read the documentation when I get the time. :wink: )

>> I also had to (decode-coding-string ... 'utf-8)

> Not sure about that, to be honest.

That surprised me too I would have assumed that since Github returns `Content-Type: text/html;charset=utf-8` `url` would do that automatically.

**Comment by tarsius** [comment]
Just in case, this is what I meant:

```lisp
(decode-coding-string
 (buffer-substring-no-properties (1+ url-http-end-of-headers) (point-max)))
 'utf-8)
```

**Comment by dgutov** [comment]
> Github returns a header with Content-Type: text/html;charset=utf-8 url would do that automatically

Indeed not. The url package is not high-level enough.

**Comment by tarsius** [comment]
Thanks again for you help! That should do it for the moment. I hope.

**Comment by tarsius** [comment]
Fixed with 1c855310909340791e5ed0410ffa6b64dbec1fed and 5648d40fa832f224e19940ba4c4f8b3837500964.

@vermiculus Currently `ghub--request` assumes that we always get back json, which isn't the case here. But this works: `(read (ghub-post "/markdown" ...))`.

**Comment by tarsius** [comment]
@vermiculus A draft is in https://github.com/tarsius/ghub/compare/read-response-function?expand=1.

Use it like:

```lisp
(let ((ghub-read-response-function 'ghub--read-raw-response))
  (ghub-post "/markdown" nil '((text . "‘guess’"))))
```

As a side-note, I am beginning to worry about people getting grossed out by all the dynamic bindings, and am considering adding the option of using function arguments instead.
